File: python/src/ancilis/dependencies/osv.py
"""OSV.dev batch vulnerability lookup — returns structured VulnerabilityFinding results."""

from __future__ import annotations

import json
import urllib.error
import urllib.request
from dataclasses import dataclass, field

from ancilis.dependencies.detector import Dependency

_OSV_BATCH_URL = "https://api.osv.dev/v1/querybatch"
_TIMEOUT = 10
_MAX_RETRIES = 2
_BATCH_SIZE = 1000


@dataclass
class VulnerabilityFinding:
    cve_id: str
    package_name: str
    installed_version: str
    severity: str  # "critical" | "high" | "medium" | "low"
    cvss_score: float | None
    fixed_version: str | None
    summary: str
    aliases: list[str] = field(default_factory=list)
    affected_versions: str = ""


def _cvss_to_severity(score: float) -> str:
    if score >= 9.0:
        return "critical"
    if score >= 7.0:
        return "high"
    if score >= 4.0:
        return "medium"
    return "low"


def _extract_cvss_score(vuln_data: dict) -> float | None:
    for sev in vuln_data.get("severity", []):
        score_type = sev.get("type", "")
        score_str = sev.get("score", "")
        if score_type in ("CVSS_V3", "CVSS_V2") and score_str:
            try:
                return float(score_str)
            except ValueError:
                pass
    return None


def _extract_severity(vuln_data: dict) -> tuple[str, float | None]:
    """Return (severity_string, cvss_score_or_None)."""
    score = _extract_cvss_score(vuln_data)
    if score is not None:
        return _cvss_to_severity(score), score
    # Fall back to database_specific.severity
    db_sev = vuln_data.get("database_specific", {}).get("severity", "")
    if db_sev:
        mapped = db_sev.lower()
        if mapped in ("critical", "high", "medium", "low"):
            return mapped, None
    return "low", None


def _extract_fixed_version(vuln_data: dict, pkg_name: str) -> str | None:
    for affected in vuln_data.get("affected", []):
        if affected.get("package", {}).get("name", "").lower() != pkg_name.lower():
            continue
        for rng in affected.get("ranges", []):
            for event in rng.get("events", []):
                fixed = event.get("fixed")
                if fixed:
                    return fixed
    return None


def _affected_summary(vuln_data: dict) -> str:
    parts: list[str] = []
    for affected in vuln_data.get("affected", []):
        for rng in affected.get("ranges", []):
            introduced = None
            fixed = None
            for event in rng.get("events", []):
                if "introduced" in event:
                    introduced = event["introduced"]
                if "fixed" in event:
                    fixed = event["fixed"]
            if introduced or fixed:
                parts.append(f">={introduced or '0'}{', <' + fixed if fixed else ''}")
    return "; ".join(parts[:3])


def _query_chunk(
    deps: list[Dependency],
    ecosystem: str = "PyPI",
) -> tuple[list[VulnerabilityFinding], str | None]:
    """Send one OSV batch request; return (findings, error_or_None)."""
    queries = [
        {
            "package": {"name": d.name, "ecosystem": ecosystem},
            "version": d.version,
        }
        for d in deps
    ]
    payload = json.dumps({"queries": queries}).encode()
    req = urllib.request.Request(
        _OSV_BATCH_URL,
        data=payload,
        headers={"Content-Type": "application/json"},
        method="POST",
    )

    last_exc: str | None = None
    for attempt in range(1, _MAX_RETRIES + 1):
        try:
            with urllib.request.urlopen(req, timeout=_TIMEOUT) as resp:
                body = json.loads(resp.read().decode())
            break
        except (urllib.error.URLError, OSError, TimeoutError) as exc:
            last_exc = str(exc)
            if attempt == _MAX_RETRIES:
                return [], last_exc
        except json.JSONDecodeError as exc:
            return [], f"Invalid JSON from OSV.dev: {exc}"
    else:
        return [], last_exc  # type: ignore[return-value]

    findings: list[VulnerabilityFinding] = []
    for dep, result in zip(deps, body.get("results", []), strict=False):
        for v in result.get("vulns", []):
            severity, cvss_score = _extract_severity(v)
            findings.append(
                VulnerabilityFinding(
                    cve_id=v.get("id", ""),
                    package_name=dep.name,
                    installed_version=dep.version or "",
                    severity=severity,
                    cvss_score=cvss_score,
                    fixed_version=_extract_fixed_version(v, dep.name),
                    summary=v.get("summary", ""),
                    aliases=v.get("aliases", []),
                    affected_versions=_affected_summary(v),
                )
            )
    return findings, None


def query_osv_batch(
    deps: list[Dependency],
    ecosystem: str = "PyPI",
) -> tuple[list[VulnerabilityFinding], str | None]:
    """Query OSV.dev for all *deps* and return (findings, error_or_None).

    Deps without a pinned version are silently skipped.
    On network failure returns ([], error_message). Non-blocking.
    """
    versioned = [d for d in deps if d.version]
    if not versioned:
        return [], None

    all_findings: list[VulnerabilityFinding] = []
    for i in range(0, len(versioned), _BATCH_SIZE):
        batch = versioned[i : i + _BATCH_SIZE]
        chunk_findings, err = _query_chunk(batch, ecosystem)
        if err is not None:
            return [], err
        all_findings.extend(chunk_findings)

    return all_findings, None

File: typescript/src/ancilis/dependencies/osv.ts
/** OSV.dev batch vulnerability lookup for npm packages. */

import type { Dependency, VulnerabilityFinding } from "./types.js";

const OSV_BATCH_URL = "https://api.osv.dev/v1/querybatch";
const TIMEOUT_MS = 10_000;
const MAX_RETRIES = 2;
const BATCH_SIZE = 1000;

type OsvSeverity = "critical" | "high" | "medium" | "low";

function cvssToSeverity(score: number): OsvSeverity {
  if (score >= 9.0) return "critical";
  if (score >= 7.0) return "high";
  if (score >= 4.0) return "medium";
  return "low";
}

function extractSeverityAndScore(
  vuln: Record<string, unknown>
): { severity: OsvSeverity; cvssScore: number | null } {
  const severities = (vuln["severity"] as Array<Record<string, unknown>>) ?? [];

  for (const sev of severities) {
    const type = sev["type"] as string | undefined;
    const scoreRaw = sev["score"] as string | number | undefined;
    if ((type === "CVSS_V3" || type === "CVSS_V2") && scoreRaw != null) {
      const score = typeof scoreRaw === "number" ? scoreRaw : parseFloat(String(scoreRaw));
      if (!isNaN(score)) {
        return { severity: cvssToSeverity(score), cvssScore: score };
      }
    }
  }

  // Fall back to database_specific severity string
  const dbSev = (
    (vuln["database_specific"] as Record<string, unknown> | undefined)?.[
      "severity"
    ] as string | undefined
  )?.toLowerCase() as OsvSeverity | undefined;

  if (dbSev && ["critical", "high", "medium", "low"].includes(dbSev)) {
    return { severity: dbSev, cvssScore: null };
  }

  return { severity: "low", cvssScore: null };
}

function extractFixedVersion(
  vuln: Record<string, unknown>,
  pkgName: string
): string | null {
  const affected = (vuln["affected"] as Array<Record<string, unknown>>) ?? [];
  for (const aff of affected) {
    const pkg = aff["package"] as Record<string, unknown> | undefined;
    if (pkg?.["name"] !== pkgName) continue;
    const ranges = (aff["ranges"] as Array<Record<string, unknown>>) ?? [];
    for (const range of ranges) {
      const events = (range["events"] as Array<Record<string, string>>) ?? [];
      for (const event of events) {
        if (event["fixed"]) return event["fixed"];
      }
    }
  }
  return null;
}

function primaryCveId(vuln: Record<string, unknown>): string {
  const id = (vuln["id"] as string) ?? "";
  const aliases = (vuln["aliases"] as string[]) ?? [];
  // Prefer CVE identifiers
  const cve = [id, ...aliases].find((a) => a.startsWith("CVE-"));
  return cve ?? id;
}

interface OsvResponse {
  results: Array<{ vulns?: Array<Record<string, unknown>> }>;
}

async function fetchWithRetry(
  body: string,
  retriesLeft: number
): Promise<OsvResponse | null> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), TIMEOUT_MS);

  try {
    const res = await fetch(OSV_BATCH_URL, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body,
      signal: controller.signal,
    });

    if (!res.ok) {
      throw new Error(`HTTP ${res.status}`);
    }

    return (await res.json()) as OsvResponse;
  } catch (err) {
    if (retriesLeft > 0) {
      return fetchWithRetry(body, retriesLeft - 1);
    }
    return null;
  } finally {
    clearTimeout(timer);
  }
}

async function queryChunk(
  deps: Dependency[]
): Promise<{ findings: VulnerabilityFinding[]; error: string | null }> {
  const queries = deps.map((dep) => ({
    package: { name: dep.name, ecosystem: "npm" },
    version: dep.version,
  }));

  const body = JSON.stringify({ queries });
  const response = await fetchWithRetry(body, MAX_RETRIES);

  if (!response) {
    return {
      findings: [],
      error: `OSV.dev unreachable after ${MAX_RETRIES + 1} attempts`,
    };
  }

  const findings: VulnerabilityFinding[] = [];
  const results = response.results ?? [];

  for (let i = 0; i < deps.length; i++) {
    const dep = deps[i]!;
    const vulns = results[i]?.vulns ?? [];

    for (const vuln of vulns) {
      const { severity, cvssScore } = extractSeverityAndScore(vuln);
      findings.push({
        cveId: primaryCveId(vuln),
        packageName: dep.name,
        installedVersion: dep.version,
        severity,
        cvssScore,
        fixedVersion: extractFixedVersion(vuln, dep.name),
        summary: (vuln["summary"] as string) ?? "",
      });
    }
  }

  return { findings, error: null };
}

export async function queryOsvBatch(
  dependencies: Dependency[]
): Promise<{ findings: VulnerabilityFinding[]; error: string | null }> {
  if (dependencies.length === 0) {
    return { findings: [], error: null };
  }

  const allFindings: VulnerabilityFinding[] = [];

  for (let i = 0; i < dependencies.length; i += BATCH_SIZE) {
    const chunk = dependencies.slice(i, i + BATCH_SIZE);
    const { findings, error } = await queryChunk(chunk);
    if (error) {
      console.warn(`[ancilis] OSV.dev lookup warning: ${error}`);
      return { findings: [], error };
    }
    allFindings.push(...findings);
  }

  return { findings: allFindings, error: null };
}
